diff -ruN vdr-2.1.1-14/menu.c vdr-2.1.1-15/menu.c
--- vdr-2.1.1-14/menu.c	2013-08-25 20:55:18.077707345 +0200
+++ vdr-2.1.1-15/menu.c	2013-08-25 20:56:31.149709855 +0200
@@ -3563,7 +3563,7 @@
 {
   Clear();
   char buffer[64];
-  snprintf(buffer, sizeof(buffer), "%s - VDR %s", tr("Setup"), VDRVERSION);
+  snprintf(buffer, sizeof(buffer), "%s - VDR %s", tr("VDR Setup"), VDRVERSION);
   SetTitle(buffer);
   SetHasHotkeys();
   Add(new cOsdItem(hk(tr("OSD")),           osUser1));
@@ -3641,9 +3641,9 @@
 :cOsdMenu("")
 {
   // Load Menu Configuration
-  cString menuXML = cString::sprintf("%s/setup/vdr-menu.%s.xml", cPlugin::ConfigDirectory(), Setup.OSDLanguage);
+  cString menuXML = cString::sprintf("%s/../menu.%s.xml", cPlugin::ConfigDirectory(), Setup.OSDLanguage);
   if (access(menuXML, 04) == -1)
-     menuXML = cString::sprintf("%s/setup/vdr-menu.xml", cPlugin::ConfigDirectory());
+     menuXML = cString::sprintf("%s/../menu.xml", cPlugin::ConfigDirectory());
   subMenu.LoadXml(menuXML);
   nrDynamicMenuEntries = 0;
   SetMenuCategory(mcMain);
@@ -3733,7 +3733,7 @@
          else if (strcmp(item, "Recordings") == 0)
             Add(new cOsdItem(hk(tr("Recordings")), osRecordings));
          else if (strcmp(item, "Setup") == 0) {
-            cString itemSetup = cString::sprintf("%s%s", tr("Setup"), *subMenu.GetMenuSuffix());
+            cString itemSetup = cString::sprintf("%s%s", tr("VDR Setup"), *subMenu.GetMenuSuffix());
             Add(new cOsdItem(hk(itemSetup), osSetup));
             }
          else if (strcmp(item, "Commands") == 0 && Commands.Count() > 0) {
diff -ruN vdr-2.1.1-14/menuitems.c vdr-2.1.1-15/menuitems.c
--- vdr-2.1.1-14/menuitems.c	2013-05-24 12:26:01.000000000 +0200
+++ vdr-2.1.1-15/menuitems.c	2013-08-25 20:56:31.149709855 +0200
@@ -1176,7 +1176,7 @@
 
 void cMenuSetupPage::SetSection(const char *Section)
 {
-  SetTitle(cString::sprintf("%s - %s", tr("Setup"), Section));
+  SetTitle(cString::sprintf("%s - %s", tr("VDR Setup"), Section));
 }
 
 eOSState cMenuSetupPage::ProcessKey(eKeys Key)
