diff -uNr vdr-softhddevice-0.6.0+ra13c25d3.orig/Makefile vdr-softhddevice-0.6.0+ra13c25d3/Makefile
--- vdr-softhddevice-0.6.0+ra13c25d3.orig/Makefile	2013-08-07 16:29:44.000000000 +0800
+++ vdr-softhddevice-0.6.0+ra13c25d3/Makefile	2013-08-11 16:50:57.032631371 +0800
@@ -18,9 +18,9 @@
     # support VDPAU video output module
 VDPAU ?= $(shell pkg-config --exists vdpau && echo 1)
     # support VA-API video output module (deprecated)
-#VAAPI ?= $(shell pkg-config --exists libva && echo 1)
+VAAPI ?= $(shell pkg-config --exists libva && echo 1)
     # support glx output
-#OPENGL ?= $(shell pkg-config --exists gl glu && echo 1)
+OPENGL ?= $(shell pkg-config --exists gl glu && echo 1)
     # screensaver disable/enable
 SCREENSAVER ?= 1
     # use ffmpeg libswresample
