diff -uNr vdr-plugin-reelchannelscan-0.6.1.orig/channelscan.c vdr-plugin-reelchannelscan-0.6.1/channelscan.c
--- vdr-plugin-reelchannelscan-0.6.1.orig/channelscan.c	2010-06-17 14:32:12.000000000 +0800
+++ vdr-plugin-reelchannelscan-0.6.1/channelscan.c	2013-07-11 08:56:51.379701812 +0800
@@ -37,6 +37,11 @@
 eAutoScanStat
     cPluginChannelscan::AutoScanStat = AssNone;
 
+#if APIVERSNUM > 10723
+cStringList
+    cPluginChannelscan::Nodes;
+#endif
+
 cScanSetup  ScanSetup;
 bool OnlyChannelList=false;
 
@@ -117,11 +122,51 @@
     return true;
 }
 
+#if APIVERSNUM > 10723
+bool
+cPluginChannelscan::GetAdapterFrontend(int CardIndex, int &Adapter, int &Frontend) {
+  Adapter = Frontend = -1;
+  if (CardIndex < 0)
+     return false;
+  if (CardIndex < Nodes.Size() &&
+      2 == sscanf(Nodes[CardIndex], "%d %d", &Adapter, &Frontend)) {
+     return true;
+  }
+  return false;
+}
+#endif
+
 bool
 cPluginChannelscan::Start(void)
 {
     // Start any background activities the plugin shall perform.
+#if (APIVERSNUM < 10507)
     RegisterI18n(Phrases);
+#endif
+#if APIVERSNUM > 10723
+    cReadDir DvbDir(DEV_DVB_BASE);
+    if (DvbDir.Ok()) {
+       struct dirent *a;
+       while ((a = DvbDir.Next()) != NULL) {
+             if (strstr(a->d_name, DEV_DVB_ADAPTER) == a->d_name) {
+                int Adapter = strtol(a->d_name + strlen(DEV_DVB_ADAPTER), NULL, 10);
+                cReadDir AdapterDir(AddDirectory(DEV_DVB_BASE, a->d_name));
+                if (AdapterDir.Ok()) {
+                   struct dirent *f;
+                   while ((f = AdapterDir.Next()) != NULL) {
+                         if (strstr(f->d_name, DEV_DVB_FRONTEND) == f->d_name) {
+                            int Frontend = strtol(f->d_name + strlen(DEV_DVB_FRONTEND), NULL, 10);
+                            Nodes.Append(strdup(cString::sprintf("%2d %2d", Adapter, Frontend)));
+                            }
+                         }
+                   }
+                }
+             }
+       }
+
+    if (Nodes.Size() > 0)
+       Nodes.Sort();
+#endif
     return true;
 }
 
diff -uNr vdr-plugin-reelchannelscan-0.6.1.orig/channelscan.h vdr-plugin-reelchannelscan-0.6.1/channelscan.h
--- vdr-plugin-reelchannelscan-0.6.1.orig/channelscan.h	2010-06-17 14:31:08.000000000 +0800
+++ vdr-plugin-reelchannelscan-0.6.1/channelscan.h	2013-07-11 08:56:51.383701812 +0800
@@ -68,6 +68,9 @@
 {
   private:
     // Add any member variables or functions you may need here.
+#if APIVERSNUM > 10723
+    static cStringList Nodes;
+#endif
   public:
     cPluginChannelscan();
     virtual ~ cPluginChannelscan();
@@ -102,6 +105,9 @@
     static eAutoScanStat AutoScanStat;
     ChannelScanData channelData_;
     bool channelDataPresent_;
+#if APIVERSNUM > 10723
+    static bool GetAdapterFrontend(int CardIndex, int &Adapter, int &Frontend);
+#endif
 };
 
 // --- cSetupMenu ------------------------------------------------------
diff -uNr vdr-plugin-reelchannelscan-0.6.1.orig/csmenu.c vdr-plugin-reelchannelscan-0.6.1/csmenu.c
--- vdr-plugin-reelchannelscan-0.6.1.orig/csmenu.c	2010-06-17 14:32:12.000000000 +0800
+++ vdr-plugin-reelchannelscan-0.6.1/csmenu.c	2013-07-11 08:56:51.391701812 +0800
@@ -243,7 +243,7 @@
                 txtstream << tr("DVB-C - Cable") << " (" << tr("Tuner") << ' ' << tuner + 1 << ')';
                 stp = CABLE;
             } else if (device->ProvidesSource(cSource::stSat)) {
-                if (device->NumProvidedSystems() == 2) { 
+                if (device->NumProvidedSystems() == 3) { 
 //                    if(TunerIsRotor(tuner))
 //                       txtstream << tr("DVB-S2 - Rotor") << " (" << tr("Tuner") << ' ' << tuner + 1 << ')';
 //                    else
@@ -378,8 +378,12 @@
 
     // avoid C/T-positions for SAT
     // and take Astra as start position
-    if (srcTypes[currentTuner] >= SAT && (source == cSource::FromString("C") || source == cSource::FromString("T"))) {
-        source = cSource::FromString("S19.2E");
+    if (srcTypes[currentTuner] >= SAT && (source == cSource::FromString("C") || source == cSource::FromString("T") || !Sources.Get(source))) {
+        cDiseqc *d = Diseqcs.First();
+        if (Setup.DiSEqC > 0 && d)
+            source = d->Source();
+        else
+            source = cSource::FromString("S19.2E");
     }
 
     Add(new cMenuEditStraItem(tr("Search Mode"), &scanMode, 2, searchTexts));
@@ -1209,7 +1213,7 @@
     {
         if (NORMALKEY(Key) == kLeft)    // TODO might want to increase the delta if repeated quickly?
         {
-           if(cSource::IsSat(source->Code()) && !cPluginManager::GetPlugin("mcli")) {
+           if(cSource::IsSat(source->Code()) && !cPluginManager::GetPlugin("mcli") && Setup.DiSEqC > 0) {
                 source = oldSrc;
                 while(!found && source && (source->Code() & cSource::stSat)) {
                     for (cDiseqc *p = Diseqcs.First(); p && !found; p = Diseqcs.Next(p)) {
@@ -1231,7 +1235,7 @@
              }
            }
         } else if (NORMALKEY(Key) == kRight) {
-            if (cSource::IsSat(source->Code()) && !cPluginManager::GetPlugin("mcli")) {
+            if (cSource::IsSat(source->Code()) && !cPluginManager::GetPlugin("mcli") && Setup.DiSEqC > 0) {
                source = oldSrc;
                while(!found && source && (source->Code() & cSource::stSat)) {
                    for (cDiseqc *p = Diseqcs.First(); p && !found; p = Diseqcs.Next(p)) {
diff -uNr vdr-plugin-reelchannelscan-0.6.1.orig/i18n.c vdr-plugin-reelchannelscan-0.6.1/i18n.c
--- vdr-plugin-reelchannelscan-0.6.1.orig/i18n.c	2010-06-17 14:31:08.000000000 +0800
+++ vdr-plugin-reelchannelscan-0.6.1/i18n.c	2013-07-11 08:56:51.395701811 +0800
@@ -23,8 +23,10 @@
  *
  ***************************************************************************/
 
+#include <vdr/config.h>
 #include "i18n.h"
 
+#if (APIVERSNUM < 10507)
 const tI18nPhrase Phrases[] = {
     {"Channel Scan",
      "Kanalsuche",
@@ -1881,3 +1883,4 @@
      },
     {NULL}
 };
+#endif
diff -uNr vdr-plugin-reelchannelscan-0.6.1.orig/i18n.h vdr-plugin-reelchannelscan-0.6.1/i18n.h
--- vdr-plugin-reelchannelscan-0.6.1.orig/i18n.h	2010-06-17 14:31:08.000000000 +0800
+++ vdr-plugin-reelchannelscan-0.6.1/i18n.h	2013-07-11 08:57:00.731701421 +0800
@@ -29,6 +29,8 @@
 #include <vdr/i18n.h>
 #include <vdr/config.h>
 
+#if (APIVERSNUM < 10507)
 extern const tI18nPhrase Phrases[];
+#endif
 
 #endif //_I18N__H
diff -uNr vdr-plugin-reelchannelscan-0.6.1.orig/Makefile vdr-plugin-reelchannelscan-0.6.1/Makefile
--- vdr-plugin-reelchannelscan-0.6.1.orig/Makefile	2010-06-17 14:32:12.000000000 +0800
+++ vdr-plugin-reelchannelscan-0.6.1/Makefile	2013-07-11 08:56:53.219701735 +0800
@@ -141,3 +141,5 @@
 	@-rm -f $(PODIR)/*.mo $(PODIR)/*.pot
 	@-rm -f $(OBJS) $(DEPFILE) *.so *.tgz core* *~
 
+install:
+	${INSTALL_PROGRAM} $(LIBDIR)/libvdr-$(PLUGIN).so.$(APIVERSION) $(PREFIX)/lib/vdr/libvdr-$(PLUGIN).so.$(APIVERSION)
diff -uNr vdr-plugin-reelchannelscan-0.6.1.orig/scan.c vdr-plugin-reelchannelscan-0.6.1/scan.c
--- vdr-plugin-reelchannelscan-0.6.1.orig/scan.c	2010-06-17 14:32:12.000000000 +0800
+++ vdr-plugin-reelchannelscan-0.6.1/scan.c	2013-07-11 08:56:53.223701735 +0800
@@ -197,7 +197,17 @@
 
 #ifndef DEVICE_ATTRIBUTES
     char buffer[265];
-    snprintf(buffer, sizeof(buffer), "/dev/dvb/adapter%d/frontend0", cardnr);
+#if APIVERSNUM > 10723
+    int Adapter = -1, Frontend = -1;
+    if (!cPluginChannelscan::GetAdapterFrontend(cDevice::GetDevice(cardnr)->CardIndex(), Adapter, Frontend)) {
+       esyslog("GetAdapterFrontend() failed");
+       cMenuChannelscan::scanState = ssDeviceFailure;
+       return false;
+    }
+    snprintf(buffer, sizeof(buffer), "/dev/dvb/adapter%d/frontend%d", Adapter, Frontend);
+#else
+    snprintf(buffer, sizeof(buffer), "/dev/dvb/adapter%d/frontend0", cDevice::GetDevice(cardnr)->CardIndex());
+#endif
 
     fd_frontend = open(buffer, O_RDONLY | O_NONBLOCK);
     if (fd_frontend <= 0) {
@@ -421,7 +431,7 @@
 void cScan::ScanDVB_S(cTransponder * tp, cChannel * c)
 {
  //const time_t tt = time(NULL);
-  int maxmods = device->NumProvidedSystems() == 2? 4 : 2;
+  int maxmods = device->NumProvidedSystems() == 3? 4 : 2;
 
    // esyslog("%s cTransponder* tp = %x  cChannel *c = %x", __PRETTY_FUNCTION__);
     esyslog("maxmods = %d",maxmods);
@@ -431,7 +441,7 @@
         ;
 
     // skip HD Transonders on SD Tuner
-    if ( !device->NumProvidedSystems() == 2 && static_cast < cSatTransponder * >(tp)->System() == 1)
+    if ( !device->NumProvidedSystems() == 3 && static_cast < cSatTransponder * >(tp)->System() == 1)
         return;
 
   unsigned  int nRadio = radioChannelNames.size();
@@ -932,12 +942,12 @@
               char *strDate;
               asprintf(&strDate,"%s", asctime(localtime(&tt)));
               strDate[strlen(strDate)-1] = 0;
-              fprintf(fp,"\n\n%s tp=%4d, %6d(%d) TV:%4d Radio:%4d in %3d sec",strDate,i , frequency, !alreadyScanned, tvChannelNames.size()-ntv, radioChannelNames.size()-nradio, (int)difftime(t_out,t_in) );
+              fprintf(fp,"\n\n%s tp=%4d, %6d(%d) TV:%4ld Radio:%4ld in %3d sec",strDate,i , frequency, !alreadyScanned, tvChannelNames.size()-ntv, radioChannelNames.size()-nradio, (int)difftime(t_out,t_in) );
               fclose(fp);
 
               fp = fopen("/tmp/tScan.log","a");
               //fprintf(fp,"\n\n%s tp=%4d, %6d/%2d/%5d TV:%4d Radio:%4d in %3dsec",strDate,i , frequency,(*tp)->Modulation(), (*tp)->Symbolrate(), tvChannelNames.size()-ntv, radioChannelNames.size()-nradio, (int)difftime(t_out,t_in) );
-              fprintf(fp,"\n\ntp=%4d, %6d/%2d/%5d TV:%4d Radio:%4d in %3dsec new:%3d",i , frequency,(*tp)->Modulation(), (*tp)->Symbolrate(), tvChannelNames.size()-ntv, radioChannelNames.size()-nradio, (int)difftime(t_out,t_in),tvChannelNames.size()-ntv+ radioChannelNames.size()-nradio );
+              fprintf(fp,"\n\ntp=%4d, %6d/%2d/%5d TV:%4ld Radio:%4ld in %3dsec new:%3ld",i , frequency,(*tp)->Modulation(), (*tp)->Symbolrate(), tvChannelNames.size()-ntv, radioChannelNames.size()-nradio, (int)difftime(t_out,t_in),tvChannelNames.size()-ntv+ radioChannelNames.size()-nradio );
               fclose(fp);
 
               free(strDate);
diff -uNr vdr-plugin-reelchannelscan-0.6.1.orig/transponders.c vdr-plugin-reelchannelscan-0.6.1/transponders.c
--- vdr-plugin-reelchannelscan-0.6.1.orig/transponders.c	2010-06-17 14:32:12.000000000 +0800
+++ vdr-plugin-reelchannelscan-0.6.1/transponders.c	2013-07-11 08:56:53.227701735 +0800
@@ -285,7 +285,7 @@
     fec_h_ = FEC_AUTO;
     fec_l_ = FEC_AUTO;
     hierarchy_ = HIERARCHY_NONE;
-    modulation_ = FE_OFDM;
+    modulation_ = QAM_AUTO;
     guard_ = GUARD_INTERVAL_AUTO;
     transmission_ = TRANSMISSION_MODE_AUTO;
 }
