diff -Nur vdr-plugin-skinpearlhd-orig/pearlhd.c vdr-plugin-skinpearlhd/pearlhd.c
--- vdr-plugin-skinpearlhd-orig/pearlhd.c	2013-06-06 22:56:34.558126864 +0200
+++ vdr-plugin-skinpearlhd/pearlhd.c	2013-06-09 16:00:47.734914884 +0200
@@ -831,6 +831,16 @@
   case mcSchedule :
   case mcChannel :
   case mcSetup :
+  case mcSetupOsd :
+  case mcSetupEpg :
+  case mcSetupDvb :
+  case mcSetupLnb :
+  case mcSetupCam :
+  case mcSetupRecord :
+  case mcSetupReplay :
+  case mcSetupMisc :
+  case mcSetupPlugins :
+  case mcPluginSetup :
   case mcTimer :
   case mcEvent :
   // Sub Menu Schedule, Channels, Setup, Edit timer
@@ -959,6 +969,16 @@
 	  case mcSchedule :
       case mcChannel :
       case mcSetup :
+      case mcSetupOsd :
+      case mcSetupEpg :
+      case mcSetupDvb :
+      case mcSetupLnb :
+      case mcSetupCam :
+      case mcSetupRecord :
+      case mcSetupReplay :
+      case mcSetupMisc :
+      case mcSetupPlugins :
+      case mcPluginSetup :
       case mcTimer :
       case mcEvent :
 	    MessageX2 = x2Menu-360;
@@ -1002,6 +1022,16 @@
   case mcSchedule :
   case mcChannel :
   case mcSetup :
+  case mcSetupOsd :
+  case mcSetupEpg :
+  case mcSetupDvb :
+  case mcSetupLnb :
+  case mcSetupCam :
+  case mcSetupRecord :
+  case mcSetupReplay :
+  case mcSetupMisc :
+  case mcSetupPlugins :
+  case mcPluginSetup :
   case mcTimer :
   case mcEvent :
     x1Item = x1Menu;
@@ -1280,6 +1310,16 @@
     case mcSchedule :
     case mcChannel :
     case mcSetup :
+    case mcSetupOsd :
+    case mcSetupEpg :
+    case mcSetupDvb :
+    case mcSetupLnb :
+    case mcSetupCam :
+    case mcSetupRecord :
+    case mcSetupReplay :
+    case mcSetupMisc :
+    case mcSetupPlugins :
+    case mcPluginSetup :
     case mcTimer :
     case mcEvent :
 	  if (PearlHDConfig.oClockText)
