diff -uNr vdr-live-0.3.0+69f84f.orig/setup.cpp vdr-live-0.3.0+69f84f/setup.cpp
--- vdr-live-0.3.0+69f84f.orig/setup.cpp	2013-05-05 04:27:09.000000000 +0800
+++ vdr-live-0.3.0+69f84f/setup.cpp	2013-09-23 18:37:36.038341969 +0800
@@ -35,7 +35,7 @@
 		m_lastChannel( 0 ),
 		m_screenshotInterval( 1000 ),
 		m_useAuth( 1 ),
-		m_adminLogin("admin"),
+		m_adminLogin("openpctv"),
 		m_channelGroups( "" ),
 		m_scheduleDuration( "8" ),
 		m_theme("marine"),
@@ -51,7 +51,7 @@
 		m_showIMDb(1),
 		m_showChannelsWithoutEPG(0)
 {
-	m_adminPasswordMD5 = "4:" + MD5Hash("live");
+	m_adminPasswordMD5 = "4:" + MD5Hash("openpctv");
 	liveplugin = cPluginManager::GetPlugin("live");
 }
 
