diff -uNr vdr-xineliboutput-1.1.0+rd7d9bf9a.orig/Makefile vdr-xineliboutput-1.1.0+rd7d9bf9a/Makefile
--- vdr-xineliboutput-1.1.0+rd7d9bf9a.orig/Makefile	2013-08-20 20:20:59.000000000 +0800
+++ vdr-xineliboutput-1.1.0+rd7d9bf9a/Makefile	2013-08-21 18:04:27.870689052 +0800
@@ -72,6 +72,7 @@
 BINDIR  ?= /usr/bin
 
 INSTALL ?= install
+DESTDIR ?= ./.install/
 
 ###
 ### Allow user defined options to overwrite defaults:
@@ -333,15 +334,15 @@
 
 $(VDRPLUGIN): $(OBJS) $(OBJS_MPG)
 	$(CXX) $(CXXFLAGS) $(LDFLAGS_SO) $(LDFLAGS) -shared $(OBJS) $(OBJS_MPG) $(LIBS) $(LIBS_VDR) -o $@
-ifeq ($(VDR_TREE), yes)
-	$(INSTALL) $@ $(LIBDIR)/
-endif
+#ifeq ($(VDR_TREE), yes)
+#	$(INSTALL) $@ $(LIBDIR)/
+#endif
 
 install-lib: $(VDRPLUGIN) $(VDRPLUGIN_SXFE) $(VDRPLUGIN_FBFE)
-ifeq ($(XINELIBOUTPUT_VDRPLUGIN), yes)
-	@echo Installing $^
-	install -D $^ $(DESTDIR)$(LIBDIR)/
-endif
+#ifeq ($(XINELIBOUTPUT_VDRPLUGIN), yes)
+#	@echo Installing $^
+#	install -D $^ $(DESTDIR)$(LIBDIR)/
+#endif
 
 install: install-lib install-i18n
 
@@ -351,9 +352,9 @@
 
 $(VDRPLUGIN_SXFE): $(OBJS_SXFE_SO)
 	$(CC) $(CFLAGS) $(LDFLAGS_SO) $(LDFLAGS) $(OBJS_SXFE_SO) $(LIBS_X11) $(LIBS_XINE) $(LIBS_JPEG) -o $@
-ifeq ($(VDR_TREE), yes)
-	$(INSTALL) $@ $(LIBDIR)/
-endif
+#ifeq ($(VDR_TREE), yes)
+#	$(INSTALL) $@ $(LIBDIR)/
+#endif
 $(VDRSXFE): $(OBJS_SXFE)
 	$(CC) $(CFLAGS) $(LDFLAGS) $(OBJS_SXFE) $(LIBS_X11) $(LIBS_XINE) $(LIBS_JPEG) $(LIBS_PTHREAD) -o $@
 
@@ -363,9 +364,9 @@
 
 $(VDRPLUGIN_FBFE): $(OBJS_FBFE_SO)
 	$(CC) $(CFLAGS) $(LDFLAGS_SO) $(LDFLAGS) $(OBJS_FBFE_SO) $(LIBS_XINE) $(LIBS_JPEG) -o $@
-ifeq ($(VDR_TREE), yes)
-	$(INSTALL) $@ $(LIBDIR)/
-endif
+#ifeq ($(VDR_TREE), yes)
+#	$(INSTALL) $@ $(LIBDIR)/
+#endif
 $(VDRFBFE): $(OBJS_FBFE)
 	$(CC) $(CFLAGS) $(LDFLAGS) $(OBJS_FBFE) $(LIBS_XINE) $(LIBS_JPEG) $(LIBS_PTHREAD) -o $@
 
