diff -uNr a/kernel/nv-acpi.c b/kernel/nv-acpi.c
--- a/kernel/nv-acpi.c	2013-11-27 09:58:00.921400347 +0800
+++ b/kernel/nv-acpi.c	2013-11-27 10:02:33.945406152 +0800
@@ -301,14 +301,14 @@
             "NVRM: nv_acpi_remove: failed to disable display switch events (%d)!\n", status);
     }
 
-    if (pNvAcpiObject->notify_handler_installed)
-    {
-        NV_ACPI_OS_WAIT_EVENTS_COMPLETE();
-
-        // remove event notifier
-        status = acpi_remove_notify_handler(device->handle, ACPI_DEVICE_NOTIFY, nv_acpi_event);
-    }
-
+    /*if (pNvAcpiObject->notify_handler_installed)
+     *{
+     *   NV_ACPI_OS_WAIT_EVENTS_COMPLETE();
+     *
+     *   // remove event notifier
+     *   status = acpi_remove_notify_handler(device->handle, ACPI_DEVICE_NOTIFY, nv_acpi_event);
+     *}
+     */
     if (pNvAcpiObject->notify_handler_installed &&
         ACPI_FAILURE(status))
     {
