diff -uNr vdr-softhddevice-0.6.0+r34fd275b42d91f8638268d551a4726eeac2c273a.orig/softhddev.c vdr-softhddevice-0.6.0+r34fd275b42d91f8638268d551a4726eeac2c273a/softhddev.c
--- vdr-softhddevice-0.6.0+r34fd275b42d91f8638268d551a4726eeac2c273a.orig/softhddev.c	2013-07-08 20:36:08.000000000 +0800
+++ vdr-softhddevice-0.6.0+r34fd275b42d91f8638268d551a4726eeac2c273a/softhddev.c	2013-07-12 17:23:44.085225927 +0800
@@ -86,7 +86,7 @@
 static volatile char NewAudioStream;	///< new audio stream
 static volatile char SkipAudio;		///< skip audio stream
 static AudioDecoder *MyAudioDecoder;	///< audio decoder
-static enum CodecID AudioCodecID;	///< current codec id
+static enum AVCodecID AudioCodecID;	///< current codec id
 static int AudioChannelID;		///< current audio channel id
 static VideoStream *AudioSyncStream;	///< video stream for audio/video sync
 
@@ -1303,8 +1303,8 @@
     VideoDecoder *Decoder;		///< video decoder
     pthread_mutex_t DecoderLockMutex;	///< video decoder lock mutex
 
-    enum CodecID CodecID;		///< current codec id
-    enum CodecID LastCodecID;		///< last codec id
+    enum AVCodecID CodecID;		///< current codec id
+    enum AVCodecID LastCodecID;		///< last codec id
 
     volatile char NewStream;		///< flag new video stream
     volatile char ClosingStream;	///< flag closing video stream
