diff -uNr a/lib/base/rawfile.cpp b/lib/base/rawfile.cpp
--- a/lib/base/rawfile.cpp	2013-07-28 16:43:53.000000000 +0800
+++ b/lib/base/rawfile.cpp	2013-12-21 20:09:12.503209760 +0800
@@ -223,8 +223,8 @@
 		return NULL;
 
 	uint8_t* p=ringBuffer->Get(Count);
-	if (Count>KILOBYTE(64))
-		Count = KILOBYTE(64);
+	if (Count>KILOBYTE(32))
+		Count = KILOBYTE(32);
 
 	if(p && Count>=TS_SIZE) {
 		if(*p!=TS_SYNC_BYTE) {
@@ -255,8 +255,8 @@
 	eSingleLocker l(m_lock);
 	int ret = 0;
 
-	while (ringBuffer->Available()<KILOBYTE(128)) {
-		ret = ringBuffer->Read(m_fd, KILOBYTE(64));
+	while (ringBuffer->Available()<KILOBYTE(64)) {
+		ret = ringBuffer->Read(m_fd, KILOBYTE(32));
 		if (ret<0)
 			break;
 	}
diff -uNr a/lib/dvb/decoder.cpp b/lib/dvb/decoder.cpp
--- a/lib/dvb/decoder.cpp	2013-06-18 00:03:17.000000000 +0800
+++ b/lib/dvb/decoder.cpp	2013-12-21 20:09:12.507209761 +0800
@@ -104,7 +104,7 @@
 		break;
 	case aDDP:
 		bypass = 0x22;
-		xine_type = 0x22;
+		xine_type = HDMV_AUDIO_84_EAC3;
 		break;
 	}
 	xineLib->setAudioType(pid, xine_type);
diff -uNr a/lib/python/Plugins/SystemPlugins/Videomode/VideoHardware.py b/lib/python/Plugins/SystemPlugins/Videomode/VideoHardware.py
--- a/lib/python/Plugins/SystemPlugins/Videomode/VideoHardware.py	2013-06-18 00:03:16.000000000 +0800
+++ b/lib/python/Plugins/SystemPlugins/Videomode/VideoHardware.py	2013-12-21 20:09:12.507209761 +0800
@@ -73,8 +73,8 @@
 		config.pc.sd_noise     = ConfigSelection(choices = {"0": _("Off"), "1": _("On")}, default="0")
 		config.pc.sd_sharpness.addNotifier(self.updateSDfeatures)
 		config.pc.sd_noise.addNotifier(self.updateSDfeatures)
-		config.pc.initial_window_width = ConfigSelection(choices = {"0": _("0"), "720": _("720"), "1280": _("1280"), "1366": _("1366"), "1920": _("1920")}, default="0")
-		config.pc.initial_window_height = ConfigSelection(choices = {"0": _("0"), "576": _("576"), "720": _("720"), "768": _("768"), "1080": _("1080")}, default="0")
+		config.pc.initial_window_width = ConfigSelection(choices = {"0": _("0"), "720": _("720"), "1280": _("1280"), "1366": _("1366"), "1600": _("1600"), "1680": _("1680"), "1920": _("1920")}, default="0")
+		config.pc.initial_window_height = ConfigSelection(choices = {"0": _("0"), "576": _("576"), "720": _("720"), "768": _("768"), "1050": _("1050"), "1080": _("1080"), "1200": _("1200")}, default="0")
 
 		# until we have the hotplug poll socket
 #		self.timer = eTimer()
