diff -uNr enigma2pc-20130727.orig/lib/dvb/cahandler.cpp enigma2pc-20130727/lib/dvb/cahandler.cpp
--- enigma2pc-20130727.orig/lib/dvb/cahandler.cpp	2013-06-18 00:03:17.000000000 +0800
+++ enigma2pc-20130727/lib/dvb/cahandler.cpp	2013-11-30 19:22:51.071878286 +0800
@@ -385,7 +385,6 @@
 			delete *it;
 			it = clients.erase(it);
 		}
-		pmtCache.clear();
 		if (pmtCache.size() > 500)
 		{
 			pmtCache.clear();
diff -uNr enigma2pc-20130727.orig/lib/python/Screens/InfoBarGenerics.py enigma2pc-20130727/lib/python/Screens/InfoBarGenerics.py
--- enigma2pc-20130727.orig/lib/python/Screens/InfoBarGenerics.py	2013-06-18 00:07:24.000000000 +0800
+++ enigma2pc-20130727/lib/python/Screens/InfoBarGenerics.py	2013-11-30 19:28:49.059885896 +0800
@@ -44,7 +44,8 @@
 from os import rename as os_rename
 import os
 from bisect import insort
-from sys import maxint
+#from sys import maxint
+maxint = 0x7FFFFFFF
 
 from RecordTimer import RecordTimerEntry, RecordTimer
 
@@ -134,7 +135,7 @@
 		self.checkUnusedTimer = eTimer()
 		self.checkUnusedTimer.callback.append(self.checkUnused)
 		self.onLayoutFinish.append(self.unhandledKeyDialog.hide)
-		eActionMap.getInstance().bindAction('', -maxint -1, self.actionA) #highest prio
+		eActionMap.getInstance().bindAction('', -maxint, self.actionA) #highest prio
 		eActionMap.getInstance().bindAction('', maxint, self.actionB) #lowest prio
 		self.flags = (1<<1)
 		self.uflags = 0
@@ -913,10 +914,10 @@
 		service = self.session.nav.getCurrentService()
 		info = service and service.info()
 		ptr = info and info.getEvent(0)
-		if ptr and ptr.getEventName() != "":
+		if ptr:
 			epglist.append(ptr)
 		ptr = info and info.getEvent(1)
-		if ptr and ptr.getEventName() != "":
+		if ptr:
 			epglist.append(ptr)
 		self.epglist = epglist
 
diff -uNr enigma2pc-20130727.orig/lib/python/Screens/ServiceInfo.py enigma2pc-20130727/lib/python/Screens/ServiceInfo.py
--- enigma2pc-20130727.orig/lib/python/Screens/ServiceInfo.py	2013-06-18 00:03:16.000000000 +0800
+++ enigma2pc-20130727/lib/python/Screens/ServiceInfo.py	2013-11-30 19:28:43.207885771 +0800
@@ -116,7 +116,7 @@
 			else:
 				aspect = "16:9"
 
-			videocodec = ("MPEG2", "MPEG4", "MPEG1", "MPEG4-II", "VC1", "VC1-SM", "-" )[self.info and self.info.getInfo(iServiceInformation.sVideoType) or -1]			
+			videocodec = ("MPEG2", "MPEG4", "MPEG1", "MPEG4-II", "VC1", "VC1-SM", "-" )[self.info and self.info.getInfo(iServiceInformation.sVideoType)]			
 
 			width = self.info and self.info.getInfo(iServiceInformation.sVideoWidth) or -1
 			height = self.info and self.info.getInfo(iServiceInformation.sVideoHeight) or -1
