diff -uNr dropbear-2013.58.orig/options.h dropbear-2013.58/options.h
--- dropbear-2013.58.orig/options.h	2013-04-18 22:58:14.000000000 +0800
+++ dropbear-2013.58/options.h	2013-09-20 19:39:06.653391604 +0800
@@ -254,7 +254,7 @@
  * OpenSSH), set the path below. If the path isn't defined, sftp will not
  * be enabled */
 #ifndef SFTPSERVER_PATH
-#define SFTPSERVER_PATH "/usr/libexec/sftp-server"
+#define SFTPSERVER_PATH "/usr/lib/sftp-server"
 #endif
 
 /* This is used by the scp binary when used as a client binary. If you're
