--- a/xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxFFmpeg.cpp	2014-01-04 16:33:30.862459415 +0800
+++ b/xbmc/cores/dvdplayer/DVDDemuxers/DVDDemuxFFmpeg.cpp	2014-01-04 16:51:04.862481229 +0800
@@ -453,7 +453,7 @@
   if (streaminfo)
   {
     /* too speed up dvd switches, only analyse very short */
-    if(m_pInput->IsStreamType(DVDSTREAM_TYPE_DVD))
+    if(m_pInput->IsStreamType(DVDSTREAM_TYPE_DVD) || m_pInput->IsStreamType(DVDSTREAM_TYPE_PVRMANAGER))
       m_pFormatContext->max_analyze_duration = 500000;
 
 
