diff -uNr u-boot-v2009.08-imx_3.0.35_4.1.0.orig/include/configs/mx6q_sabresd.h u-boot-v2009.08-imx_3.0.35_4.1.0/include/configs/mx6q_sabresd.h
--- u-boot-v2009.08-imx_3.0.35_4.1.0.orig/include/configs/mx6q_sabresd.h	2013-11-15 23:18:43.214269379 +0800
+++ u-boot-v2009.08-imx_3.0.35_4.1.0/include/configs/mx6q_sabresd.h	2013-11-15 23:23:51.906256470 +0800
@@ -126,12 +126,12 @@
 			"ip=dhcp nfsroot=${serverip}:${nfsroot},v3,tcp\0"\
 		"bootcmd_net=run bootargs_base bootargs_nfs; "		\
 			"tftpboot ${loadaddr} ${kernel}; bootm\0"	\
-		"bootargs_mmc=setenv bootargs ${bootargs} ip=dhcp "     \
-			"root=/dev/mmcblk0p1 rootwait\0"                \
+		"bootargs_mmc=setenv bootargs ${bootargs} "     \
+			"root=/dev/mmcblk0p1 rootwait video=mxcfb1:dev=ldb,LDB-XGA,if=RGB666 video=mxcfb0:dev=hdmi,1920x1080M@60 video=mxcfb1:off video=mxcfb2:off fbmem=28M consoleblank=0\0"                \
 		"bootcmd_mmc=run bootargs_base bootargs_mmc; "   \
 		"mmc dev 3; "	\
 		"mmc read ${loadaddr} 0x800 0x2000; bootm\0"	\
-		"bootcmd=run bootcmd_net\0"                             \
+		"bootcmd=run bootcmd_mmc\0"                             \
 
 
 #define CONFIG_ARP_TIMEOUT	200UL
